### ISSUE TITLE
Using getUserMedia to create a PWA to solve LensLok

### ISSUE THREAD

**Issue by omgmog** [remotehack/remotehack.github.io#96]
Anyway this is something I've been thinking about. I've started a basic vue.js project and setup some crappy js to grab a snapshot from a camera feed to a canvas byte array 🤷‍♂️ 

I'll see how much further I can take it during the day.

The plan is basically to use the device camera to unscramble lenslok screens

process:
- get image using getUserMedia
- split in to fixed number of columns (11 I think going by the diagram), 
- codify shuffle/flip per column
- ?????
- profit!

![image](https://user-images.githubusercontent.com/50949/105728823-7ae19e80-5f24-11eb-91d6-ffcfa36f5a70.png)
![image](https://user-images.githubusercontent.com/50949/105728858-85039d00-5f24-11eb-8037-9d9654ed63a3.png)
![image](https://user-images.githubusercontent.com/50949/105728923-9482e600-5f24-11eb-8e15-1b526fb3b74b.png)

Info:
- [https://k1.spdns.de/Vintage/Sinclair/82/Peripherals/Lenslok/](https://k1.spdns.de/Vintage/Sinclair/82/Peripherals/Lenslok/)
- [http://www.computinghistory.org.uk/det/56946/Lenslok-Software-Protection-Device/](http://www.computinghistory.org.uk/det/56946/Lenslok-Software-Protection-Device/)
- [http://www.crashonline.org.uk/24/images/lenslok_diagrams.jpg](http://www.crashonline.org.uk/24/images/lenslok_diagrams.jpg)
- [http://www.cpcwiki.eu/index.php/Lenslok](http://www.cpcwiki.eu/index.php/Lenslok)
- [https://torrentfreak.com/crazy-video-game-drm-prism-1980s-style-080617/](https://torrentfreak.com/crazy-video-game-drm-prism-1980s-style-080617/)

**Comment by benfoxall** [comment]
This would be SO COOL.

* if your canvas byte array function could go both ways,  would you be able to generate lenslok graphics/videos too?
* [drawImage](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage) might be a nifty way to write video segments straight to a canvas?  Or you could use a shaderrrr!? 🔥 
* you could make a "hyper secure" video chat by outputting lens-locked video so someone would need your app to "decode" it.

**Comment by omgmog** [comment]
Yeah in the proof of concept I bashed together last week I can dump the currently streaming `getUserMedia` from a `<video>` element to `<canvas>` using `drawImage`, and then I just need to work out how to manipulate the pixels obtained by `getImageData` to do things. 


So far I have managed to flipp all of the pixel data with some example function I grabbed:

```javascript
    // flip an imageData object
    const flipImageData = (canvas) => {
      var context = canvas.getContext('2d');
      var w = canvas.width;
      var h = canvas.height;
      var imDat = context.getImageData(0,0,w,h);
      var mirImDat = context.createImageData(imDat);   // a. initialze mirImDat; all components of mirImDat.data are 0 here
      for (var y = 0; y < h; y++) {                    // b. update the mirImDat.data components
        for (var x = 0; x < w; x++) {
          var d = 4 * (y * w + x);
          var s = 4 * (y * w + (w - (x + 1)));
          mirImDat.data [d + 0] = imDat.data [s + 0];
          mirImDat.data [d + 1] = imDat.data [s + 1];
          mirImDat.data [d + 2] = imDat.data [s + 2];
          mirImDat.data [d + 3] = imDat.data [s + 3];
        };                                              // done updating the mirImDat.data components
      };
      return mirImDat;
    }
```

so I'll just need to work out how to apply that to discrete column ranges

I also need to work out how to re-order ranges, and also have a calibration step, good UX, etc. etc.

Should keep me busy on Saturday!

**Comment by benfoxall** [comment]
Ah, nice!

Yeah, I was saying that you could avoid pixel manipulation all together by drawing columns of the image straight to the canvas in regions,  which _should_ keep things on the gpu.

For instance if you wanted to flip the right/left side of a video:

```js
ctx.drawImage(video, 
                0, 0,
                video.videoWidth / 2, video.videoHeight, 
                canvas.width / 2, 0, 
                canvas.width / 2, canvas.height
)

ctx.drawImage(video, 
                video.videoWidth / 2, 0, 
                video.videoWidth / 2, video.videoHeight, 
                0, 0, 
                canvas.width / 2, canvas.height
)
```

Pico-8 would be suchhhhh a cool place to print them!

**Comment by omgmog** [comment]
I'll pick this up again some other time!
